Update package name regex for deb

LeonidVas opened the pull request:
Before the patch, it is assumed that the deb package name format is
\<package\>\_\<version\>.\<revision\>-\<dist\>\_\<arch\>.deb.
But when trying to use this formatt, we have several problems:
#16
#14

So now we aasume the deb package name format is
\<Package\>\_\<VersionNumber\>-\<DebianRevisionNumber\>\_\<DebianArchitecture\>.deb
according to https://www.debian.org/doc/manuals/debian-faq/pkg-basics.en.html#pkgname

Closes #16

knazarov commented:
I don’t remember exactly why I did this, but I suppose I might have looked at other packages, not only on tarantool. Did you try to verify this change on a sample of packages from debian/centos?

LeonidVas commented:
> I don’t remember exactly why I did this, but I suppose I might have looked at other packages, not only on tarantool. Did you try to verify this change on a sample of packages from debian/centos?

This patch only applies to deb repositories. I looked at the package name format from the documentation. I tried to find repos with a format where the distribution name would be indicated, but I could not find it. I met a similar format (which was before the patch) somewhere on StackOverflow, but could not find where it comes from. If you know deb repos with a similar name format, I would take a look at them.

LeonidVas commented:
It seems logical to me to support the format of the name described in the documentation if we have no reason to support the additional format.

Totktonada commented:
NB: For different package types: https://www.debian.org/doc/manuals/debian-reference/ch02.en.html#_debian_package_file_names (via https://github.com/tarantool/cartridge-cli/issues/303).

Totktonada commented on debrepo.py:
```diff
@@ -256,28 +256,34 @@ def dump_string(self):
+    # Since we are not using the "debian_revision" information, we do not need
+    # to separate it from the "upstream_version".
+    # https://www.debian.org/doc/debian-policy/ch-controlfields.html#version
+    expr = r'^(?P<package>[^_]+)_(?P<version>[0-9]([A-Za-z0-9.+~-])*)?_(?P<arch>[^\.]+)\.deb$'
```
It does not corresponds [2.5.8. Debian package file names](https://www.debian.org/doc/manuals/debian-reference/ch02.en.html#_debian_package_file_names) sections of the Debian manual.

\<upstream-version\> is not optional. No colon (`:`) in \<upstream-version\> characters. Optional parts are omitted together with a preceding delimiter (as I see in the [actual data](http://mirror.yandex.ru/debian/pool/main/t/tarantool/)). \<package-name\> allowed characters does not correspond to the table in the docs (not everything except underscore (`_`) is allowed).

If you're going to stick with the document, please, don't do it partially: it'll create extra confusion.

And I would implement a fair parsing with \<upstream-version\> and \<debian.version\> in order to ease comparison between the document and the regexp.

I'll also write several words about testing below.

LeonidVas replied:
Updated.

Totktonada commented:
Regarding testing: I propose to crawl all package file names for a recent Debian version (Bullseye) and for a recent LTS Ubuntu (Focal) and create a unit test for the `split_pkg_path()` function based on this data. You can use, say, https://mirror.yandex.ru. We can generate a `unittest` based test and commit it to the repository.

So we all will be sure that we're doing thing right and will stop tweak this regexp (I hope).

knazarov commented:
I've added a few unit tests in #35 and set up CI to run them automatically.

The situation is not very pretty with debian package naming. The names are really ambiguous. For example:

```
pool/main/t/tarantool/tarantool-python_0.6.6-21_amd64.deb
```
or 

```
pool/main/t/tarantool/tarantool-python_0.6.6-deb9u1_amd64.deb
```

In the first scenario, `21` is most likely a version, while in the second `deb9u1` is obviously (at least for a human reader) a dist version. There is really no way to disambiguate those two properly.

What I propose to do here is if there is ambiguity in what the last component before `_<arch>` is, then treat it as part of the version if it is a dash followed by regular number. If it has anything other than numbers, then it is a `dist`.

After your change, all the tests in #35 should pass when uncommented.

Totktonada commented:
BTW, I would feel more comfortable to have repository data placed as data, not code. Like [here](https://github.com/tarantool/multivac/tree/4a886bb2c0a775db5e274c6d01e7ebbe43b92f71/test/sensors).

LeonidVas commented:
> BTW, I would feel more comfortable to have repository data placed as data, not code. Like [here](https://github.com/tarantool/multivac/tree/4a886bb2c0a775db5e274c6d01e7ebbe43b92f71/test/sensors).

Why? in the current implementation it's more comfortable to work with it. We don't need to parse some txt file and we don't need to deal with paths. What are the benefits of your approach?

Totktonada commented:
I have no strict points here, so I don't insist. My reasons are:

1. I don't know limitations of the interpreter.
2. It is not code, but data. So .yaml feels more appropriate
3. `cloc` will exclude it from Python code statistics.

LeonidVas commented:
> I have no strict points here, so I don't insist. My reasons are:
> 
> 1. I don't know limitations of the interpreter.
> 2. It is not code, but data. So .yaml feels more appropriate
> 3. `cloc` will exclude it from Python code statistics.

I like the third one. I updated the code and moved the data with the package names into separate text files.

Totktonada reviewed (approved):
LGTM.
